---
title: "RCAS Generation and Concentration-Response Profiling"
author: "Jesse Rogers"
date: "December 6, 2023"
output: 
  html_document:
    toc: true
    toc_depth: 4
    toc_float: true
    number_sections: true
    df_print: paged
    code_folding: show
---

Extraction and analyis of Reference Class Associated Signatures (RCAS)
using U-2 OS and HepaRG chemical screening data. RCAS represent
unique gene sets more potent for individual classes of chemicals versus
those annotated for other modes-of-action.

Genes are selected by comparing benchmark doses (BMDs) as potency estimates
across classes of reference chemicals assigned to similar modes-of-action, and
genes uniquely potent for individual reference classes are selected as part of a
signature for that class:

![](Fig1A.png)

# Setup Environment

```{r env-setup, message=FALSE}
# import packages for visualization
# (note that other required packages will be imported by R scripts)
library(ggplot2)
library(ComplexHeatmap)
library(RColorBrewer)
library(circlize)

# set directories (`dir_home` should be set to the top-level directory)
dir_home <- "~/NAM-integration/NAM-integration-pilot-packaged/"
dir_scripts <- paste0(dir_home, "scripts/")

# import scripts for RCAS generation and profiling
source(paste0(dir_scripts, "pipeline_refchem_assignment.R"))
source(paste0(dir_scripts, "pipeline_RCAS_generation.R"))
source(paste0(dir_scripts, "pipeline_RCAS_profiling.R"))

path_rcas <- c(
    paste0(dir_home, "data/examples/rcas_heparg_gene.RData"),
    paste0(dir_home, "data/examples/rcas_u2os_gene.RData")
)
```

# Assign Reference Chemicals

Reference classes are developed using
[RefChemDB](https://doi.org/10.1186/s13040-022-00292-z) by
assigning chemicals to clusters of target-mode combinations. Chemicals are
assigned using a rule-based procedure:

-   Hierarchically cluster target_mode entries based on Jaccard distance 
    -   Determine clusters using a dendrogram height cutoff of 0.8 
-   For each chemical, count the number of target_mode entries that belong to 
each cluster 
-   Assign chemical to cluster with highest count 
    -   Settle ties using total support for each cluster (sum(support) for all 
target_mode entries in a cluster) 
    -   Remove remaining ties (equal target representation and support) 
-   Remove clusters with <3 chemicals assigned
-   Cluster labels assigned based on most frequent target_mode entries within
each cluster

```{r assign-refchems}
# assign reference chemicals from RefChemDB and export to file
path_refchems <- paste0(dir_home, "data/examples/refchemdb_clusters.RData")
if (!file.exists(path_refchems)) {
  refchem_assign <- assignRefChems(path_refchems)
} else {
  load(path_refchems)
}

# summarize chemicals represented by clusters
refchem_assign %>%
    count(cluster, cluster_target) %>%
    arrange(desc(n)) %>%
    DT::datatable()
```

# Generate RCAS

Gene sets are generated for reference classes using High-Throughput
Transcriptomics (HTTr) screening data of 1,201 chemicals in U-2 OS and 
HepaRG cell lines. Gene-level potency estimates were previously generated from
DESeq2-moderated fold changes using the `httrpathway` package
[available on USEPA Github](https://github.com/USEPA/CompTox-httrpathway), and 
are used as the basis for determining gene sets.

## Conduct univariate potency analysis

`analyzeHTTrANOVA()` can take some time to run if many significant
genes are found (compiling HepaRG-based RCAS below takes ~10 minutes). Here, 
RCAS objects are pre-generated and saved in the `data/examples` directory.

```{r generate-rcas, eval=FALSE}
# specify paths for httr gene-level potency estimates
path_httr <- c(
  "/ccte/projects1/HTTr/screen_gene_cr/GENE_CR_heparg2d_toxcast_pfas_pe1_normal_0.05_conthits.RData",
  "/ccte/projects1/HTTr/screen_gene_cr/GENE_CR_u2os_toxcast_pfas_pe1_normal_0.05_conthits.RData"
)

# generate RCAS for each dataset
rcas_heparg <- analyzeHTTrANOVA(
    filepath.data = path_httr[1],
    filepath.ref = path_refchems,
    col.httr.type = "gene",
    col.ref.class = "cluster_target",
    metric.fill = 2.5
)
save(rcas_heparg, file = path_rcas[1])

rcas_u2os <- analyzeHTTrANOVA(
    filepath.data = path_httr[2],
    filepath.ref = path_refchems,
    col.httr.type = "gene",
    col.ref.class = "cluster_target",
    metric.fill = 2.5
)
save(rcas_u2os, file = path_rcas[2])

# display results for significant genes: U-2 OS
rcas_u2os$posthoc.estimate$posthoc.sig %>%
    mutate(across(where(is.numeric), ~ signif(.x, 3))) %>%
    DT::datatable()
```

## Select final RCAS genes

Following ANOVA-based potency comparisons per gene, final gene sets are
constructed for each reference class by selecting genes whose BMD values are
significantly lower for a single class versus at least 70% of other classes.
RCAS are retained for each target if at least 10 genes meet this criterion.

Note that this selection step is performed within `profileRCAS()` and is done
here for visualizing RCAS sizes/potencies.

```{r compile-rcas}
# perform RCAS selection for each object
composite_heparg <- selectRCASGenes(path_rcas[1])
composite_u2os <- selectRCASGenes(path_rcas[2])

# combine rcas for each cell type
composite_all <- rbind(
    mutate(
        composite_heparg$composite,
        source = "heparg",
        n_classes = length(unique(composite_heparg$httr.wide$ref_class))
    ),
    mutate(
        composite_u2os$composite,
        source = "u2os",
        n_classes = length(unique(composite_u2os$httr.wide$ref_class))
    )
)

# compare sizes of retained RCAS
composite_all %>%
    count(source, class_1) %>%
    rename(class = class_1, size = n) %>%
    DT::datatable()
```

## Compare RCAS gene-level potency estimates

```{r ht-fcn, class.source = 'fold-hide'}
plotHeatmap <- function(rcas, title, composite, width = 8, height = 8) {
  if (is.data.frame(composite) & title %in% composite$source) {
    rcas.vars <- composite %>%
      filter(source == title) %>%
      pull(variable)
    rcas.vars <- gsub("\`", "", rcas.vars)
    rcas.varclass <- composite %>%
      filter(source == title) %>%
      pull(class_1)
    # get classes/chems with passing variables
    rcas.classes.unique <- composite %>%
      filter(source == title) %>%
      pull(class_1) %>%
      unique()
    rcas.names <- rcas$httr.wide %>%
      filter(ref_class %in% rcas.classes.unique) %>%
      pull(name)
    rcas.classes <- rcas$httr.wide %>%
      filter(ref_class %in% rcas.classes.unique) %>%
      pull(ref_class)
    # filter log(bmd) data for variables + chems
    rcas.mat <- rcas$httr.wide %>%
      filter(ref_class %in% rcas.classes.unique) %>%
      select(any_of(rcas.vars)) %>%
      as.matrix()
    rownames(rcas.mat) <- rcas.names
  } else {
    rcas.vars <- rcas$posthoc.estimate$posthoc.vars
    rcas.vars <- gsub("\`", "", rcas.vars)

    rcas.mat <- rcas$httr.wide %>%
      select(any_of(rcas.vars)) %>%
      as.matrix()
    rcas.classes.unique <- unique(rcas$httr.wide$ref_class)
    rcas.names <- rcas$httr.wide$name
    rcas.classes <- rcas$httr.wide$ref_class
    rownames(rcas.mat) <- stringr::str_trunc(rcas.names, 20)
  }
  # construct annotation color palette
  rcas.palette <- colorRampPalette(
      brewer.pal(8, "Set2")
    )(length(unique(composite$class_1)))
  names(rcas.palette) <- unique(composite$class_1)
  # plot heatmap
  rcas.ht <- Heatmap(
    rcas.mat,
    colorRamp2(c(-2.5, -1.25, 0, 1.25, 2.5), rev(brewer.pal(5, "Spectral"))),
    name = paste0("log10(BMD)"),
    column_title = paste0("RCAS-identified genes: ", gsub("_", " ", title)),
    column_title_gp = gpar(fontsize = 8),
    row_title = "Reference Chemical",
    row_title_gp = gpar(fontsize = 8),
    left_annotation = rowAnnotation(
      reference_class = rcas.classes,
      col = list(reference_class = rcas.palette),
      show_annotation_name = FALSE,
      annotation_legend_param = list(
        reference_class = list(direction = "horizontal", nrow = 2)
      ),
      annotation_width = unit(0.2, "cm")
    ),
    top_annotation = HeatmapAnnotation(
      variable_class = rcas.varclass,
      col = list(variable_class = rcas.palette),
      show_annotation_name = FALSE,
      show_legend = FALSE,
      annotation_height = unit(0.2, "cm")
    ),
    row_split = rcas.classes,
    column_labels = rep("", ncol(rcas.mat)),
    row_names_gp = gpar(fontsize = 7),
    heatmap_legend_param = list(
      direction = "horizontal",
      at = seq(-3, 3, 1)
    ),
    width = unit(width, "cm"),
    height = unit(height, "cm")
  )
  return(rcas.ht)
}
```

```{r compare-gene-potency}
# rename select refchems to common names
rename_dtxsids <- c("DTXSID8023474", "DTXSID6040745", "DTXSID9040760")
rename_names <- c("Pimozide", "L-165041", "GW0742")
for (idx in seq_len(length(rename_dtxsids))) {
  composite_heparg$httr.wide$name[composite_heparg$httr.wide$dtxsid == rename_dtxsids[idx]] <- rename_names[idx]
  composite_u2os$httr.wide$name[composite_u2os$httr.wide$dtxsid == rename_dtxsids[idx]] <- rename_names[idx]
}
# plot heatmaps
ht_heparg <- plotHeatmap(
  composite_heparg, "heparg", composite_all, width = 6, height = 6
)
draw(ht_heparg, merge_legend = TRUE, heatmap_legend_side = "bottom")
ht_u2os <- plotHeatmap(
  composite_u2os, "u2os", composite_all, width = 6, height = 4.8
)
draw(ht_u2os, merge_legend = TRUE, heatmap_legend_side = "bottom")
```

# Profile RCAS

Resulting RCAS for each reference class are subsequently used to profile
targeted bioactivitiy across all 1,201 chemicals screened in U-2 OS/HepaRG
cell lines. Signature-level profiling is conducted using `httrpathway`, which
performs ssGSEA  ([Barbie et al. 2009](https://doi.org/10.1038/nature08460))
for each chemical-dose group for each RCAS and subsequently conducts
concentration profiling for each chemical and RCAS via the `tcplfit2` package
([Sheffield et al 2022](https://doi.org/10.1093/bioinformatics/btab779)).

Notes:

-   Current scripts assume access to the HTTr internal EPA mongodb for accessing log2(fold-change) data. Future versions of this data may be hosted on Clowder as flat files and available for public use.
  -   For mongodb access, user credentials shold be stored in user's home directory under `~/.mongopw`, following a json structure per study.
  -   Current scripts also assume installation of the `httrlib` package (currently available for internal use only). Alternatively, these functions can be sourced from the httrpathway cloned repository using `source("../httrpathway/httrpl/Rlib/httrpl.R", chdir = TRUE)`.
-   **Import and processing of log2(fold-change) data will take a long time (1-1.5 hours per dataset)**

```{r profile-rcas, eval=FALSE}
# define mongodb connection and export args
db_host <- "ccte-mongodb-res.epa.gov"
db_names <- c(
  "res_httr_heparg2d_toxcast",
  "res_httr_u2os_toxcast"
)
fc1_names <- c(
  "heparg_fc1_rcas.RData",
  "u2os_fc1_rcas.RData"
)
catalog_names = c(
  "signatureDB_rcas_heparg.xlsx",
  "signatureDB_rcas_u2os.xlsx"
)

# run RCAS profiling per cell type
profileRCAS(
  filepath_rcas = path_rcas[1],
  db_host = db_host,
  db_name = db_names[1],
  fc1_name = fc1_names[1],
  random_sigs = TRUE,
  catalog_name = catalog_names[1]
)
profileRCAS(
  filepath_rcas = path_rcas[2],
  db_host = db_host,
  db_name = db_names[2],
  fc1_name = fc1_names[2],
  random_sigs = TRUE,
  catalog_name = catalog_names[2]
)
```

## Compare RCAS signature-level potency estimates

Signature-level curve fits are extracted for all chemicals used in RCAS
generation for comparison of potency/efficacy estimates. Estimates are 
compared between in-class chemicals (i.e. reference class matches RCAS) and
out-of-class chemicals (i.e. reference class does not match RCAS) as validation
of performance of signature-level profiling against the reference set.

```{r compare-signature-potency}
# import RCAS signature results
dir_cr <- "../output/signature_conc_resp_summary/"
db_names <- c(
  "res_httr_heparg2d_toxcast",
  "res_httr_u2os_toxcast"
)
files_cr <- paste0("SIGNATURE_CR_rcas_", db_names, "_gsea_0.05_conthits.RData")
load(paste0(dir_cr, files_cr[1]))
cr_heparg <- SIGNATURE_CR
load(paste0(dir_cr, files_cr[2]))
cr_u2os <- SIGNATURE_CR
rm(SIGNATURE_CR)

# combine results across cell types + filter for reference chemicals
cr_all <- rbind(
  mutate(cr_heparg, source = "heparg"),
  mutate(cr_u2os, source = "u2os")
)
cr_refchems <- rbind(
  filter(mutate(cr_heparg, source = "heparg"), dtxsid %in% composite_heparg$ref$dtxsid),
  filter(mutate(cr_u2os, source = "u2os"), dtxsid %in% composite_u2os$ref$dtxsid)
) %>%
  mutate(ref_class = composite_heparg$ref$ref_class[
    match(dtxsid, composite_heparg$ref$dtxsid)
  ])


# visualize potency/efficacy for each class
for (idx in seq_len(length(rename_dtxsids))) {
  cr_refchems$name[cr_refchems$dtxsid == rename_dtxsids[idx]] <- rename_names[idx]
}
plot_pot_eff <- cr_refchems %>%
  filter(!grepl("Random", signature)) %>%
  mutate(
    bmd_log = case_when(!is.na(bmd) ~ log10(bmd), TRUE ~ 2.5),
    toc_signed = top_over_cutoff * sign(top),
    pass_bioactivity = hitcall >= 0.9 & top_over_cutoff >= 1.5,
    source_signature = paste0(source, "_", signature),
    in_class = ref_class == signature,
  ) %>%
  ggplot(aes(
      bmd_log, toc_signed,
      color = in_class,
      shape = pass_bioactivity,
      label = stringr::str_trunc(name, 30)
  )) +
  geom_point(alpha = 0.9) +
  geom_hline(yintercept = 0, color = "grey50") +
  geom_hline(
      yintercept = c(-1.5, 1.5),
      linetype = "dashed", color = "grey50"
  ) +
  scale_shape_manual(values = c(3, 16)) +
  scale_color_brewer(palette = "Set1") +
  facet_wrap(~source_signature) +
  labs(
      x = "log10(BMD)", y = "Signed Top Over Cutoff",
      color = "Chemical in\nreference\nclass",
      shape = "Passed\nbioactivity\nthresholds",
      title = ""
  ) +
  theme_bw() +
  theme(legend.direction = "vertical")

plot_pot_eff
```

## Conduct randomization test



# Reproducibility

```{r reproducibility}
sessionInfo()
```